packages/agreement-process/src/services/agreementService.ts

```diff
@@ -741,3 +764,76 @@ export async function addConsumerDocumentLogic(
     agreement.metadata.version
   );
 }
+
+export async function rejectAgreementLogic({
+  agreementId,
+  rejectionReason,
+  authData,
+  agreementQuery,
+  tenantQuery,
+  eserviceQuery,
+}: {
+  agreementId: string;
+  rejectionReason: string;
+  authData: AuthData;
+  agreementQuery: AgreementQuery;
+  tenantQuery: TenantQuery;
+  eserviceQuery: EserviceQuery;
+}): Promise<CreateEvent<AgreementEvent>> {
+  const agreementToBeRejected = await agreementQuery.getAgreementById(
+    agreementId
+  );
+  assertAgreementExist(agreementId, agreementToBeRejected);
+
+  assertRequesterIsProducer(agreementToBeRejected.data, authData);
+
+  assertExpectedState(agreementId, agreementToBeRejected.data.state, [
+    agreementState.pending,
+  ]);
+
+  const eservice = await eserviceQuery.getEServiceById(
+    agreementToBeRejected.data.eserviceId
+  );
+  assertEServiceExist(agreementToBeRejected.data.eserviceId, eservice);
+
+  const consumer = await tenantQuery.getTenantById(
+    agreementToBeRejected.data.consumerId
+  );
+  assertTenantExist(agreementToBeRejected.data.consumerId, consumer);
+
+  const descriptor = eservice.data.descriptors.find(
+    (d) => d.id === agreementToBeRejected.data.descriptorId
+  );
+  if (descriptor === undefined) {
```

Comment:
From find documentation, this method return `undefined` if item not found
> find(predicate: (value: T, index: number, obj: T[]) => unknown, thisArg?: any): T | undefined;

```suggestion
  if (!descriptor) {
```

💡  also...  
this assertion could be a new assertion to add in `validators.ts` ?